In [1]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
import pandas as pd
from sklearn.metrics import *

In [2]:
'''   와인 분류하기
1 Decision Tree 사용해 보기
2 Random Forest 사용해 보기
3 SVM 사용해 보기
4 SGD Classifier 사용해 보기
5 Logistic Regression 사용해 보기 ''' 

'   와인 분류하기\n1 Decision Tree 사용해 보기\n2 Random Forest 사용해 보기\n3 SVM 사용해 보기\n4 SGD Classifier 사용해 보기\n5 Logistic Regression 사용해 보기 '

In [3]:
# (2) 데이터 준비, (3) 데이터 이해하기
wine = load_wine()
wine_data = wine.data
wine_label = wine.target
wine_df = pd.DataFrame(data=wine_data, columns=wine.feature_names)
wine_df


alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  
0                            3.92   1065.0  
1                            3.40   1050.0  
2                            3.17   1185.0  
3                            3.45   1480.0  
4                            2.93    735.0  
..                            ...      ...  
173                          1.74    740.0  
174                          1.56    750.0  
175                          1.56    835.0  
176                          1.62    840.0  
177                          1.60    560.0  

[178 rows x 13 columns]

In [5]:
wine_df["label"] = wine.target
wine_df

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  label  
0                            3.92   1065.0      0  
1                            3.40   1050.0      0  
2                            3.17   1185.0      0  
3                            3.45   1480.0      0  
4                            2.93    735.0      0  
..                            ...      ...    ...  
173                          1.74    740.0      2  
174                          1.56    750.0      2  
175                          1.56    835.0      2  
176                          1.62    840.0      2  
177                          1.60    560.0      2  

[178 rows x 14 columns]

In [7]:
print(wine_data.shape)
wine.target_names

(178, 13)


array(['class_0', 'class_1', 'class_2'], dtype='<U7')

In [8]:

wine_df["label"].value_counts() 

1    71
0    59
2    48
Name: label, dtype: int64

In [9]:
# (4) train, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(wine_data, 
                                                    wine_label, 
                                                    test_size=0.2, 
                                                    random_state=10)


In [13]:
from sklearn.tree import DecisionTreeClassifier

# (5) 모델 학습 및 예측 
# (5) - 1 Decision Tree 
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

accuracy = accuracy_score(y_test, y_pred)
accuracy

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        10
           1       1.00      0.89      0.94        18
           2       0.80      1.00      0.89         8

    accuracy                           0.94        36
   macro avg       0.93      0.96      0.94        36
weighted avg       0.96      0.94      0.95        36



0.9444444444444444

In [26]:
# (5) - 2 Random Forest
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))


              precision    recall  f1-score   support

           0       1.00      1.00      1.00        10
           1       1.00      0.89      0.94        18
           2       0.80      1.00      0.89         8

    accuracy                           0.94        36
   macro avg       0.93      0.96      0.94        36
weighted avg       0.96      0.94      0.95        36



In [21]:
from sklearn import svm

svm_model = svm.SVC(random_state=32)
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))


              precision    recall  f1-score   support

           0       0.88      0.70      0.78        10
           1       0.81      0.72      0.76        18
           2       0.42      0.62      0.50         8

    accuracy                           0.69        36
   macro avg       0.70      0.68      0.68        36
weighted avg       0.74      0.69      0.71        36



In [25]:
from sklearn.linear_model import SGDClassifier

sgd_model = SGDClassifier(random_state=32)
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred, zero_division = 0))
#label2의 경우는 전혀 positive로 감지하지 않아 0값이 나오면서 오류가 발생한 부분을 해결하였다.


              precision    recall  f1-score   support

           0       0.62      0.80      0.70        10
           1       0.70      0.89      0.78        18
           2       0.00      0.00      0.00         8

    accuracy                           0.67        36
   macro avg       0.44      0.56      0.49        36
weighted avg       0.52      0.67      0.58        36



In [22]:
from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression(random_state=32, solver = 'liblinear')
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))


              precision    recall  f1-score   support

           0       1.00      1.00      1.00        10
           1       1.00      0.83      0.91        18
           2       0.73      1.00      0.84         8

    accuracy                           0.92        36
   macro avg       0.91      0.94      0.92        36
weighted avg       0.94      0.92      0.92        36



## 와인 데이터 평가 지표

- 잘못된 제품으로 판별하면 안되기 때문에 FP와 FN 값을 줄여야 하며, 이 때 중요한 평가 지표로는 precision과 recall이 있다. 두 지표 값을 모두 고려할 수 있는 f1-score를 평가 지표로 사용해야한다.
- 데이터가 label2 표본이 적은 편인데, 대략적으로 48/178 label2가 아닐 확률이 73%나 된다. SGD Classifier 모델의 경우 2번을 아예 고르지도 않아서 precision, recall, f1score 모두 0이 나왔는데 이런 오차 속에서도 정확도가 0.67이나 나온다.

결론은 RandomForestClassifier 또는 DecisionTreeClassifier를 사용해야한다.